### PR TITLE
feature/webapi

### PR DESCRIPTION
**Pull request by KOKAProduktion**
Hiho @albar965, hiho @u-an-i,

here's an implementation of 2 new API routes, hopefully matching @u-an-i requirements:

- fixes for some issues mentioned during https://github.com/albar965/littlenavmap/pull/731#pullrequestreview-733971322
- sim aircraft info (position, speed, altitude)
- map painters zoom values (int value)

See the [API contract](https://github.com/KOKAProduktion/littlenavmap/blob/feature/webapi/web/webapi.yaml) for details 

**Note**: I had to make `WebMapController` and its `MapPainterWidget` accessible via `NavApp` in order to get the proper map zoom value. May i ask you to particularly take a look at this if there's nothing wrong with doing it like that?
Commit: https://github.com/albar965/littlenavmap/commit/453aaebd419cecb83b412319484039296b12301c

Thanks in advance for reviewing and hear you soon!
Cheers!

### CODE REVIEW

File: src/webapi/uiactionscontroller.cpp [hunk 1]
```diff
@@ -25,7 +28,8 @@ WebApiResponse UiActionsController::infoAction(WebApiRequest request){
     WebApiResponse response = getResponse();
 
     UiInfoData data = {
-        getMainWindow()->getMapWidget()->zoom()
+        getMainWindow()->getMapWidget()->zoom(),
+        getNavApp()->getWebController()->getWebMapController()->getMapPaintWidget()->zoom()
```
Comment (KOKAProduktion) [inline src/webapi/uiactionscontroller.cpp:32]: This is the line that Irequired for web ui map zoom value retrieval. I had to create `WebController.getWebMapController()`and `WebMapController.getMapPaintWidget()` for it

**Comment by u-an-i** [comment]
i skimmed over the changes, partly slower, partly faster, I didn't spot any issue. But a lot to digest calmly. How about a textual representation of the relation between the classes, a UML lite, in form of a readme? (like i attempted here for the web plugin interface: https://github.com/albar965/littlenavmap/pull/767#issuecomment-912794842)

**Comment by u-an-i** [comment]
actually, I found https://github.com/KOKAProduktion/littlenavmap/wiki/Little-Navmap-Web-API-Documentation again :)

**Comment by albar965** [comment]
@u-an-i and @KOKAProduktion 
I quickly skimmed over it but I'll leave the code reviews mostly for you guys now. I have a big pile of issues here: https://github.com/albar965/littlenavmap/milestone/11 Not all will make it into 2.8.

Please let me know before you take an issue or start any other tasks to avoid double work.

**Comment by u-an-i** [comment]
I was busy with https://github.com/albar965/littlenavmap/pull/773 . I hope I can take a detailed look this week.

File: src/common/abstractinfobuilder.h [hunk 2]
```diff
@@ -90,8 +108,24 @@ class AbstractInfoBuilder : public QObject
...
-
+  /**
+   * @brief get pretty printed coords from Pos
+   * @param rec
+   * @return printable string
+   */
+  virtual QString getLocalizedCoordinatesString(const Pos& pos) const;
+  /**
+   * @brief get a map of coordinate data from suitable sql record
+   * @param pos
```
Comment (u-an-i) [inline src/common/abstractinfobuilder.h:119]: mixup of doc : rec

File: src/common/abstractinfobuilder.h [hunk 3]
```diff
@@ -90,8 +108,24 @@ class AbstractInfoBuilder : public QObject
    * @return printable string
    */
   virtual QString getCoordinatesString(const Pos& pos) const;
-
-
+  /**
+   * @brief get pretty printed coords from Pos
+   * @param rec
```
Comment (u-an-i) [inline src/common/abstractinfobuilder.h:113]: mixup of doc: rec

File: web/webapi.yaml [hunk 4]
```diff
@@ -64,33 +102,33 @@ components:
           type: object
           properties:
             ATIS:
-              type: string
-              description: Only set if available
-              example: "123,450 MHz"
+              type: number
+              description: Only set if available. (Format 000,000 MHz)
+              example: "123450"
```
Comment (u-an-i) [inline web/webapi.yaml:107]: does this mean the unit is actually Hz? Perhaps add the unit the returned value/current example is in to the description (i read you are supposed to deliver raw values).

**Comment by u-an-i** [comment]
i read through this PR. looks good, 3 minor comments.

File: web/webapi.yaml (hunk 4 above)
Comment (albar965) [reply web/webapi.yaml:107]: I'd either use Hz as unit or add an additional field `kHz` or `MHz`.

**Comment by albar965** [comment]
Stupid question: Can you update PRs with changes or do you have to create a new one? Never seen this from the other side.
Just asking in case I merged too early.

**Comment by u-an-i** [comment]
i never bothered to read up on this either and wondered myself :) :( :|  I think the PR is gone but if the branch didn't get deleted a new PR can be created from it and git might recognize all other now unchanged but originally new/changed files in such a PR are now no change in comparison to the target branch.

File: web/webapi.yaml (hunk 4 above)
Comment (KOKAProduktion) [reply web/webapi.yaml:107]: *Facepalm! Obviously it's in Hz..., yeah! I'll fix that! Thx!

**Comment by u-an-i** [comment]
btw: when using my plugin https://github.com/albar965/littlenavmap/pull/775, the debug shows:

```
[2021-09-10 17:15:48.764 DEBUG]: virtual bool atools::fs::sc::SimConnectHandler::connect() Before Open
[2021-09-10 17:15:51.025 DEBUG]: HttpRequest: from 2a01:c23:c1cd:f000:18c3:f987:d1ac:3c17: GET /api/sim/info HTTP/1.1
[2021-09-10 17:15:53.792 DEBUG]: virtual bool atools::fs::sc::SimConnectHandler::connect() Before Open
```

Idk whether the "connect before open" is part of your calls.

**Comment by KOKAProduktion** [comment]
Hiho @albar965, hiho @u-an-i,

thank you @u-an-i for your review time and comments! 

>  Can you update PRs with changes or do you have to create a new one? 

@albar965 : I have pushed changes to my repo's branch which did not appear in the PR after it was closed, so yes, i guess changes to a PR are only possible before the merge. So in our case it might have been too soon :)

Now i have a fix for providing the maps `distance` instead of the `zoom` value for which it seems i have to create a new PR for.

If you're fine with it, @u-an-i could request a merge from you after a successful review?

Thx and Cheers!

**Comment by u-an-i** [comment]
i am fine with this.

**Comment by KOKAProduktion** [comment]
> 
> 
> btw: when using my plugin #775, the debug shows:
> 
> ```
> [2021-09-10 17:15:48.764 DEBUG]: virtual bool atools::fs::sc::SimConnectHandler::connect() Before Open
> [2021-09-10 17:15:51.025 DEBUG]: HttpRequest: from 2a01:c23:c1cd:f000:18c3:f987:d1ac:3c17: GET /api/sim/info HTTP/1.1
> [2021-09-10 17:15:53.792 DEBUG]: virtual bool atools::fs::sc::SimConnectHandler::connect() Before Open
> ```
> 
> Idk whether the "connect before open" is part of your calls.

Hmmm, looks like init output of `SimConnectHandler` which I didn't look into very much :) Although it's most certainly called by my `SimActionsController` API controller.

Don't know if that's a problem?

File: web/webapi.yaml (hunk 4 above)
Comment (albar965) [reply web/webapi.yaml:107]: Be careful where you take this stuff from. :smile: `map::MapVor` uses MHz * 1000 and `map::MapNdb` kHz * 100.

**Comment by u-an-i** [comment]
> Don't know if that's a problem?

neither do i know :)  but my build of lnm doesn't connect to FS2020. Perhaps this is relevant. Does yours connect?

**Comment by KOKAProduktion** [comment]
> 
> 
> > Don't know if that's a problem?
> 
> neither do i know :) but my build of lnm doesn't connect to FS2020. Perhaps this is relevant. Does yours connect?

Yes, it's successfully passing through the SimConnect data from FS2020 (Win10) on my host. But i remember i had to recompile LNM with SimConnect support first. Now that you mention it, i recall having some issues before it finally worked...

Also I don't have any other flight sim to test it with...

**Comment by u-an-i** [comment]
> recall having some issues with that before it finally worked...

i have the issue "simconnect.h" not found when i set the environment variable, thus i out-commented that variable.


> Do you run this from a deploy target?

yes, and both files are present

**Comment by KOKAProduktion** [comment]
> Do you run this from a deploy target? There should be a `SimConnect.dll` and `littlenavmap.exe.simconnect` in the LNM CWD.
> This should normally not be a problem. Otherwise I'd have an army with pitchforks in front of the the house.

Yeah, it's running directly from Qt Creator as `Release` target with SimConnect support. I should've noted it down, but i think i had to copy the .dll's from the released LNM version folder... Or something like that. Sorry! I can investigate a bit further if you like.

**Comment by u-an-i** [comment]
i try some things and report back, hang on a few minutes if you like :)

**Comment by albar965** [comment]
> > btw: when using my plugin #775, the debug shows:
> > ```
> > [2021-09-10 17:15:48.764 DEBUG]: virtual bool atools::fs::sc::SimConnectHandler::connect() Before Open
> > [2021-09-10 17:15:51.025 DEBUG]: HttpRequest: from 2a01:c23:c1cd:f000:18c3:f987:d1ac:3c17: GET /api/sim/info HTTP/1.1
> > [2021-09-10 17:15:53.792 DEBUG]: virtual bool atools::fs::sc::SimConnectHandler::connect() Before Open
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Idk whether the "connect before open" is part of your calls.
> 
> Hmmm, looks like init output of `SimConnectHandler` which I didn't look into very much :) Although it's most certainly called by my `SimActionsController` API controller.
> 
> Don't know if that's a problem?

Why do you deal with the SimConnectHandler? You only need to get the SimConnectUserAircraft from the NavApp. The rest happens in the background.

**Comment by albar965** [comment]
> i have the issue "simconnect.h" not found when i set the environment variable, thus i out-commented that variable.

Path with spaces?

**Comment by u-an-i** [comment]
> Path with spaces?

yes 🥇  :|   I let rebuild now ...

**Comment by KOKAProduktion** [comment]
> Why do you deal with the SimConnectHandler?

I don't. At least not intentionally :)

https://github.com/albar965/littlenavmap/blob/9d40ae62d038f91f034d7ff8f8d87a889de44dbd/src/webapi/abstractlnmactionscontroller.cpp#L150-L152

> i have the issue "simconnect.h" not found when i set the environment variable, thus i out-commented that variable.

Have you checked both atools and lnm project files for their `SIMCONNECT_PATH`s

LNM:
https://github.com/albar965/littlenavmap/blob/9d40ae62d038f91f034d7ff8f8d87a889de44dbd/littlenavmap.pro#L96
ATOOLS:
https://github.com/albar965/atools/blob/f06f732c11cade8a7cda4811b80639692702a19d/atools.pro#L95

For atools i had to adjust it like this 

```
INCLUDEPATH += $$SIMCONNECT_PATH"\inc"
->
INCLUDEPATH += $$SIMCONNECT_PATH"\include"
```

**Comment by albar965** [comment]
inc vs. include: Probably because I use an ancient SC from FSX. This is to keep LNM with old stuff compatible.
Sorry about the problem with the spaces in the path. Quoting paths in qmake and Windows is a real PITA.

**Comment by albar965** [comment]
> @albar965 : I have pushed changes to my repo's branch which did not appear in the PR after it was closed, so yes, i guess changes to a PR are only possible before the merge. So in our case it might have been too soon :)

Meh. :frowning_face:  Sorry, give me a ping once a PR is done.

**Comment by KOKAProduktion** [comment]
> 
> 
> inc vs. include: Probably because I use an ancient SC from FSX. This is to keep LNM with old stuff compatible.
> Sorry about the problem with the spaces in the path. Quoting paths in qmake and Windows is a real PITA.

It's still not working? But I'd say that it must be doable if it works with the released version, no?

I'm afraid fake data wouldn't help your development either, would it?

**Comment by u-an-i** [comment]
> Sorry about the problem with the spaces in the path

i thought i should have detected this ,  no worry for you :)

atools now compiled with simconnect support using "include" rather than "inc".

Thank you both!

**Comment by albar965** [comment]
> > inc vs. include: Probably because I use an ancient SC from FSX. This is to keep LNM with old stuff compatible.
> > Sorry about the problem with the spaces in the path. Quoting paths in qmake and Windows is a real PITA.
> 
> It's still not working? But I'd say that it must be doable if it works with the released version, no?
> 
> I'm afraid fake data wouldn't help your development either, would it?

I don't understand. The release version does not use spaced paths. 
Fake data?

**Comment by KOKAProduktion** [comment]
> 
> 
> > > inc vs. include: Probably because I use an ancient SC from FSX. This is to keep LNM with old stuff compatible.
> > > Sorry about the problem with the spaces in the path. Quoting paths in qmake and Windows is a real PITA.
> > 
> > 
> > It's still not working? But I'd say that it must be doable if it works with the released version, no?
> > I'm afraid fake data wouldn't help your development either, would it?
> 
> I don't understand. The release version does not use spaced paths.
> Fake data?

Never mind. I've misread that :) Glad it works!

**Comment by albar965** [comment]
Correction: I use `ATOOLS_SIMCONNECT_PATH=C:\Program Files (x86)\Microsoft Games\Microsoft Flight Simulator X SDK\SDK\Core Utilities Kit\SimConnect SDK`. No problems there.

**Comment by albar965** [comment]
The SimConnect dummy is needed to build the Linux and macOS versions if this is what you mean by fake data.

**Comment by KOKAProduktion** [comment]
> 
> 
> The SimConnect dummy is needed to build the Linux and macOS versions if this is what you mean by fake data.

Nope, i thought about the API returning dummy sim data for @u-an-i auto-zoom feature development if simconnect would'nt work for him. But that may be feasible to forge in order to be useful... Was just an idea.

**Comment by u-an-i** [comment]
simconnect now works for me, i can now test and optimise my autozoom plugin ! :)

Thank you again! (@KOKAProduktion if you like to have a look at no status code being returned by the web api (so a fetch response.ok is false) you can do so (/api/sim/info), otherwise you can wait til i set my autozoom PR to "not draft".

After that I look into your new PR.

**Comment by albar965** [comment]
> > The SimConnect dummy is needed to build the Linux and macOS versions if this is what you mean by fake data.
> 
> Nope, i thought about the API returning dummy sim data for @u-an-i auto-zoom feature development if simconnect would'nt work for him. But that may be feasible to forge in order to be useful... Was just an idea.

I see.

There is more dummy stuff maybe useful to test auto-zoom for you guys: 
Add `DEFINES += DEBUG_MOVING_AIRPLANE` in `build_options.pro` (included from all `.pro` files from `..`), load a flight plan and you can move a dummy aircraft with `Shift+Ctrl` and mouse movement to simulate a moving aircraft. The altitude of the aircraft follows the elevation profile.

Another one: 
Start `Little Navconnect` with `replay-gui` and you can record and replay flights from the menu. No UI to scroll forth and back, though.

**Pull request merged**